### DISCUSSION TITLE
Morph-To-Many and MorphTo

### DISCUSSION THREAD

**Discussion by nilskretschmer** [staudenmeir/eloquent-has-many-deep#228] [Q&A] [answered]
Hi,

thank you @staudenmeir for this (another great) package!

Currently I am trying to get my head around a complex polymorphic structure. I am extending the awesome https://github.com/spatie/laravel-permission package to have roles/permissions also attached to polymorphic resources.

Examples:
- User A is Admin of Resource A (in this example a project resource - could be something else hence polymorphic)
- User B has permissions to edit resource A (in this example also a project resource - could be something else hence polymorphic)

I have the following structure:

`users:`
| id | 
| -- | 

`projects:`
| id |
| -- |

`roles:`
| id | model_id | model_type |
| -- | ---------- | ------------- |

*The model_id and model_type in this case relate to a resource (model type other than user)*

`permissions:`
| id | model_id | model_type |
| -- | ---------- | ------------- |

*The model_id and model_type in this case relate to a resource (model type other than user)*

`model_has_roles:`
| model_id | model_type | role_id |
| ---------- | ------------- | ------- |

*The model_id and model_type in this case relate to a user model  (default or other user model)*

`model_has_permissions:`
| model_id | model_type | permission_id |
| ---------- | ------------- | --------------- |

*The model_id and model_type in this case relate to a user model  (default or other user model)*

`role_has_permissions:`
| permission_id | role_id |
| --------------- | ------- |

*The model_id and model_type in this case relate to a user model  (default or other user model)*

Now I want to get all projects a user is associated with by this "double" polymorphic relationship. Like:

`Project.php`
```
class Project extends Model
{
    public function roles(): MorphMany
    {
        return $this->MorphMany(Role::class, 'model');
    }

    public function users(): ? // how to get users through roles

}
```

`Role.php`
```
class Role extends SpatieRole
{
    public function model(): MorphTo
    {
        return $this->morphTo();
    }
    
    // from the spatie package role class
    public function users(): BelongsToMany
    {
        return $this->morphedByMany(
            getModelForGuard($this->attributes['guard_name'] ?? config('auth.defaults.guard')),
            'model',
            config('permission.table_names.model_has_roles'),
            app(PermissionRegistrar::class)->pivotRole,
            config('permission.column_names.model_morph_key')
        );
    }
}
```

`User.php`
```
class User extends Authenticatable
{
    public function projects(): ? // get projects through roles here
}
```

I am not sure if I really need this package but it seems like so. But I did not find examples in the docs that match my use case. Can you help me out on this?

I was also looking in to your https://github.com/staudenmeir/eloquent-json-relations package. But I am not sure if this is suitable in my use case but it looks also promising. Especially by eliminating a lot of pivot tables in my case..

Thank you in advance!

**Comment by staudenmeir** [comment]
Hi @nilskretschmer,

> Now I want to get all projects a user is associated with by this "double" polymorphic relationship. Like:

Is this the path the relationship needs to "travel"?

`users` -> `model_has_roles` -> `roles` -> `projects`

**Reply by nilskretschmer** [reply]
Hi @staudenmeir 

Yes.

It is like:

`users` -> morphMany (intermediate table `model_has_roles`) -> `roles` -> morphTo -> `projects`

**Comment by staudenmeir** [comment, marked as answer]
You can define the `users()` relationship from your existing relationships:

```php
class Project extends Model
{
    use \Staudenmeir\EloquentHasManyDeep\HasRelationships;

    public function users(): HasManyDeep
    {
        return $this->hasManyDeepFromRelations($this->roles(), (new Role())->users());
    }
}
```

Then use it to define the reverse relationship:

```php
class User extends Authenticatable
{
    use \Staudenmeir\EloquentHasManyDeep\HasRelationships;

    public function projects(): HasManyDeep
    {
        return $this->hasManyDeepFromReverse((new Project)->users());
    }
}
```

**Reply by nilskretschmer** [reply]
Ah nice! I was not aware that I can use existing relationships when polymorphic relations are involved. I will give it a try thank you!
